pkg/cmd/pulumi/convert/convert.go

```diff
@@ -457,13 +482,153 @@ func runConvert(
+) (map[string]*schema.PackageDescriptor, hcl.Diagnostics, error) {
+	// TODO use the package block to get the packages and parameterize them and generate SDKs.
+	files, err := os.ReadDir(sourceDirectory)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not read sourceDirectory: %w", err)
+	}
+
+	parser := hclsyntax.NewParser()
+	_, err = pcl.ParseFiles(parser, sourceDirectory, files)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("could not parse pcl file: %w", err)
+	}
+
+	allPackageDescriptors := make(map[string]*schema.PackageDescriptor)
+
+	var diagnostics hcl.Diagnostics
+	for _, file := range parser.Files {
+		packageDescriptors, diags := pcl.ReadPackageDescriptors(file)
+		diagnostics = append(diagnostics, diags...)
+		for packageName, descriptor := range packageDescriptors {
+			if _, ok := allPackageDescriptors[packageName]; ok {
+				diagnostics = append(diagnostics, errorf(file.Body.Range(), "package %q was already defined", packageName))
+				continue
+			}
+			allPackageDescriptors[packageName] = descriptor
+		}
+	}
+
+	if len(diagnostics) != 0 {
+		var errorDiags hcl.Diagnostics
+		for _, d := range diagnostics {
+			if d.Severity == hcl.DiagError {
+				errorDiags = append(errorDiags, d)
+			}
+		}
+
+		if len(errorDiags) != 0 {
+			return nil, diagnostics, nil
+		}
+	}
+
+	return allPackageDescriptors, nil, nil
+}
+
+func generateAndLinkSdksForPackages(
+	ctx context.Context,
+	ws pkgWorkspace.Context,
+	language string,
+	sdkTargetDirectory string,
+	convertOutputDirectory string,
+	pkgs map[string]*schema.PackageDescriptor,
+) error {
+	for _, pkg := range pkgs {
+		tempOut, err := os.MkdirTemp("", "gen-sdk-for-dependency-")
+		if err != nil {
+			return fmt.Errorf("failed to create temporary directory: %w", err)
+		}
+
+		var parameterizationValue []byte
+		if pkg.Parameterization != nil {
+			parameterizationValue = pkg.Parameterization.Value
+		} else {
+			// Only generate SDKs for packages that have parameterization for now, others should be implicit.
+			continue
+		}
+
+		pkgSchema, err := pcmd.SchemaFromSchemaSourceValueArgs(ctx, pkg.Name, parameterizationValue)
+		if err != nil {
+			return fmt.Errorf("creating package schema: %w", err)
+		}
+
+		err = pcmd.GenSDK(
+			language,
+			tempOut,
+			pkgSchema,
+			/*overlays*/ "",
+			/*local*/ true,
+		)
+		if err != nil {
+			return fmt.Errorf("error generating sdk: %w", err)
+		}
+
+		sdkOut := filepath.Join(sdkTargetDirectory, pkg.Parameterization.Name)
+		err = pcmd.CopyAll(sdkOut, filepath.Join(tempOut, language))
+		if err != nil {
+			return fmt.Errorf("failed to move SDK to project: %w", err)
+		}
+
+		err = os.RemoveAll(tempOut)
+		if err != nil {
+			return fmt.Errorf("could not remove temp dir: %w", err)
+		}
+
+		fmt.Printf("Generated local SDK for package '%s:%s'\n", pkg.Name, pkg.Parameterization.Name)
+
+		// Change working directory to the newly generated project output so
+		// workspace instance behaves.
```

Comment:
I think for comments like this, it can be valuable to explain how things would misbehave if we didn't do it -- as it stands this leaves me with more questions than answers.